pkg/reslife/lifecycle.go

```diff
@@ -0,0 +1,256 @@
+	}
+	Purger interface {
+		Purge(ctx context.Context) error
+	}
+	LifeCycleerFactory func(ctx context.Context, config cfg.Config, logger log.Logger) (LifeCycleer, error)
+	managerCtxKey      struct{}
+)
+
+type Settings struct {
+	Create struct {
+		Enabled bool `cfg:"enabled" default:"false"`
+	} `cfg:"create"`
+	Init struct {
+		Enabled bool `cfg:"enabled" default:"true"`
+	} `cfg:"init"`
+	Register struct {
+		Enabled bool `cfg:"enabled" default:"true"`
+	} `cfg:"register"`
+	Purge struct {
+		Enabled bool `cfg:"enabled" default:"false"`
+	} `cfg:"purge"`
+}
+
+type LifeCycleManager struct {
+	logger     log.Logger
+	clock      clock.Clock
+	settings   *Settings
+	refresher  func(fac LifeCycleerFactory) (LifeCycleer, error)
+	resources  []LifeCycleer
+	created    []string
+	registered []string
+	purged     []string
+}
+
+func ProvideLifeCycleManager(ctx context.Context, config cfg.Config, logger log.Logger) (*LifeCycleManager, error) {
+	return appctx.Provide(ctx, managerCtxKey{}, func() (*LifeCycleManager, error) {
+		logger = logger.WithChannel("lifecycle-manager")
+
+		settings := &Settings{}
+		config.UnmarshalKey("resource_lifecycles", settings)
+
+		return &LifeCycleManager{
+			logger:   logger,
+			clock:    clock.Provider,
+			settings: settings,
+			refresher: func(fac LifeCycleerFactory) (LifeCycleer, error) {
+				return fac(ctx, config, logger)
+			},
+			resources:  nil,
+			created:    []string{},
+			registered: []string{},
+			purged:     []string{},
+		}, nil
+	})
+}
+
+func (m *LifeCycleManager) refreshResources(ctx context.Context) error {
+	var err error
+	var container *[]LifeCycleerFactory
+	var resource LifeCycleer
+
+	if container, err = provideContainer(ctx); err != nil {
+		return fmt.Errorf("could not get lifecyleers: %w", err)
+	}
+
+	for _, factory := range *container {
+		if resource, err = m.refresher(factory); err != nil {
+			return fmt.Errorf("could not build lifecycleer: %w", err)
+		}
+
+		m.resources = append(m.resources, resource)
+	}
+
+	return nil
+}
+
+func (m *LifeCycleManager) Create(ctx context.Context) error {
+	var ok bool
+	var creator Creator
+
+	if !m.settings.Create.Enabled {
+		m.logger.Info("create lifecycle not enabled, skipping")
+
+		return nil
+	}
+
+	if err := m.refreshResources(ctx); err != nil {
+		return fmt.Errorf("could not refresh resources: %w", err)
+	}
+
+	for _, res := range m.resources {
+		if funk.Contains(m.created, res.GetId()) {
+			continue
+		}
+
+		now := m.clock.Now()
+
+		if creator, ok = res.(Creator); !ok {
+			continue
+		}
+
+		if err := creator.Create(ctx); err != nil {
+			return fmt.Errorf("could not create resource %q: %w", res.GetId(), err)
+		}
+
+		m.created = append(m.created, res.GetId())
+
+		took := m.clock.Since(now)
+		m.logger.Info("created resource %s in %s", res.GetId(), took)
+	}
+
+	m.logger.Info("executed the create lifecycle")
+
+	return nil
+}
+
+func (m *LifeCycleManager) Init(ctx context.Context) error {
```

Comment:
Why is init special?
- it doesn't check if something was already initialized
- there is not time measured
- there is no log that this happened